### ISSUE TITLE
Unselect candidates after they've been acted on?

### ISSUE THREAD

**Issue by localauthor** [oantolin/embark#629]
I'm playing with the new `embark-select` feature, for the purpose of integrating it with some of my packages. When acting on selected candidates in a buffer, the candidates stay selected after they've been acted on. What do you think about adding an option to unselect candidates after they have been acted on---something like `embark-quit-after-action`?

**Comment by oantolin** [comment]
This seems like a reasonable option to add, I only hesitate because Embark already has so many options.

For what it's worth, dired doesn't unmark files after you run a command on them either.

**Comment by localauthor** [comment]
Two fair points. No worries then. ;)

**Comment by oantolin** [comment]
Well, I'm certainly open to further discussion on this. As I said, I think it is a reasonable thing to want.

**Comment by localauthor** [comment]
On the issue of unselecting candidates, and perhaps a way forward with this: there might be a reason to have a post-action-hook that clears `embark--selection` for certain actions. For example, when using `delete-region` on candidates --- the regions are gone but the candidates are still in `embark--selection`, and even the overlays seem to remain.

**Comment by oantolin** [comment]
I've actually made us of the fact that the selection remains after `delete-region`: after `delete-region` you can still `embark-insert` the deleted text to bring it together, for example. But I agree it feels a little weird to have the selection preserved when it is no longer present in the buffer.

**Comment by localauthor** [comment]
This is a bit of a problem from a UX perspective, since there is no visual indicator that there are still selections in the buffer. So if you later select other candidates in the same buffer, you end up with more candidates than you might expect.

**Comment by oantolin** [comment]
Yeah, I've been thinking that maybe I should add an indicator (this is a technical term: the functions that go in `embark-indicators`) saying how many items are selected.

**Comment by localauthor** [comment]
That could certainly be helpful. How best to go about it, another question entirely :)

**Comment by minad** [comment]
I think it is okay to keep the candidates selected, in order to act again for example. It is also easy to unmark all via `C-. A SPC`. What about adding an Embark selected indicator to the modeline?

**Comment by minad** [comment]
Better variant:

~~~ elisp
(advice-add #'embark-select :after (lambda (&rest _) (force-mode-line-update t)))

(defun embark--selection-mode-line ()
  (when-let ((sel
              (buffer-local-value
               'embark--selection
               (or (when-let ((win (active-minibuffer-window)))
                     (window-buffer win))
                   (current-buffer)))))
    (format #("  Embark:%s " 1 12 (face (embark-selected bold))) (length sel))))

(push '(:eval (embark--selection-mode-line)) mode-line-misc-info)
~~~

**Comment by oantolin** [comment]
I'm not sure taking up space in the modeline is warranted. Don't you only need to know whether there is a selection or not when you call `embark-act`? That's why I was suggesting an Embark indicator.

**Comment by minad** [comment]
I think showing this information always is warranted since the selection is some state which is there even if you are not inside the `embark-act` context. But of course this should be configurable and entirely optional. We could guard this with an option and only if the option is non-nil add `(:eval (embark--selected-mode-line))` to `mode-line-misc-info`. Modifying `mode-line-misc-info` could happen inside of `embark-select`.

**Comment by minad** [comment]
`embark--selected-mode-line` is maybe simple enough for the wiki or only of wiki-quality anyway. In my setup I try to keep as few custom additions as possible, so if it would ship with Embark I would probably use this mode line indicator. Also note that the Vertico selection indication is independent of `embark-act` and always shown.

**Comment by oantolin** [comment]
OK, that sounds reasonable, but I think the mode line indicator should be wrapped up in a minor mode, no? Automatically messing with people's mode lines is a sure-fire way to make enemies!

**Comment by minad** [comment]
> OK, that sounds reasonable, but I think the mode line indicator should be wrapped up in a minor mode, no? Automatically messing with people's mode lines is a sure-fire way to make enemies!

A minor mode is too much. I would simply make this an option. If the option is enabled, the mode line is modified and otherwise it is not. Also note that the modification will only happen if you select something. I will prepare a PR and we can discuss the indicator question separately.

> And further thinking about this issue, I think maybe a new user option embark-deselect-after-action is reasonable.

Agree.

**Comment by localauthor** [comment]
> @localauthor do you have any further experience with this issue? Do you still find manually deselecting to be too tedious?

No further experience to report, since I did a little post-action-hook of my own 😜
But I would appreciate the alist option you describe!

**Comment by localauthor** [comment]
Here's the function I'm using in the hook, btw:
```elisp
(defun embark--clear-selection (&rest _)
    "Clear buffer-local candidates selected with `embark-select'."
    (mapc
     (lambda (x) (delete-overlay (cdr x)))
     embark--selection)
    (setq-local embark--selection nil))
```

**Comment by oantolin** [comment]
Which actions do you use that as a post-action hook for, @localauthor?

**Comment by oantolin** [comment]
@minad With the current information that hooks get, a hook can't know whether you acted on the selection or not. So currently, the best you could do with hooks is deselect unconditionally, whether or not you acted on the selection (like @locaauthor's hook). I don't think that solution is good enough because it turns selections into a sort of modal interface: you need to act on the selection before you act on anything else in the buffer.

Obviously we could add a user option `embark-deselect-after-action`, but we could instead pass an extra argument to hooks indicating whether the selection was acted on.

**Comment by minad** [comment]
Hmm, thinking about this more I wonder if maybe a post action hook is not the correct abstraction. Acting on a selection is triggered by embark-act-all while post action hooks run per action. Omar, what do you think?

**Comment by oantolin** [comment]
> Hmm, thinking about this more I wonder if maybe a post action hook is not the correct abstraction. Acting on a selection is triggered by embark-act-all while post action hooks run per action. Omar, what do you think?

A post-action hook could still work if it knew that you had acted on the selection. It would have to handle two cases still: whether you used a multitarget action or a one-target-at-a-time action.

It would be easier to use a user option, since then you don't need to handle those two cases separately.

**Comment by minad** [comment]
> A post-action hook could still work if it knew that you had acted on the selection. It would have to handle two cases still: whether you used a multitarget action or a one-target-at-a-time action.

Yes. But the hook would be executed repeatedly, right?

> It would be easier to use a user option, since then you don't need to handle those two cases separately.

Seems like it. Or a more general post-act-all hook, if we would discover more use cases which need to run after acting on all candidates.

**Comment by oantolin** [comment]
Another issue is whether we'd want collect and export to also clear the selection. I think we probably would. But those commands can be called directly, without going through `embark-act-all` (which now that I think of it, is kind of weird, they probably should be regarded as multitarget-actions). So they need to be handled specially. 

I think we do have an `embark-after-export-hook` which could be used to clear the selection, but that takes a different kind of hook, one called without parameters, and it is called from the exported buffer. So we'd need to write a second hook which switched back to the buffer the selection came from and then clears the selection.

**Comment by minad** [comment]
Okay, I see. And we would still need the additional keyword argument telling the hook if we are operating in a selection or not, right? When operating on a selection the hook would do the deselection otherwise it would do nothing.

Hard to tell if this leads to more code and more confusion in comparison to an extra deselect-action option.

**Comment by oantolin** [comment]
Right, I think the hooks are a little more complicated. Even a hypothetical post-act-all hook as you say would need to know whether the selection was acted on or not: in Embark's default configuration the selection is the highest priority candidate collector, but a user might add in their personal configuration another candidate collector with higher priority and it would awkward if their selection got cleared anyway!

**Comment by minad** [comment]
> Even a hypothetical post-act-all hook as you say would need to know whether the selection was acted on or not: in Embark's default configuration the selection is the highest priority candidate collector, but a user might add in their personal configuration another candidate collector with higher priority and it would awkward if their selection got cleared anyway!

Hmm, I would ignore that problem and just assume that selections always have the highest priority.

Also, what about deselecting in Dired and Ibuffer? Maybe the candidates collector should return a deselect function in addition to the candidates if deselection is applicable.

**Comment by oantolin** [comment]
> Also, what about deselecting in Dired and Ibuffer? Maybe the candidates collector should return a deselect function in addition to the candidates

Well, normal dired commands don't deselect, so it would feel a little strange to configure embark to deselect your dired selection for you. I think this deselection discussion should be limited to the embark selection. And maybe the default should be to not deselect, to follow dired's lead.

**Comment by oantolin** [comment]
Maybe `:candidates` is a little suspicious, maybe in the multitarget case I should still call the hook once per target, even though the action takes them all at once in a list argument...

**Comment by minad** [comment]
I just remembered that we have some special handling for embark--restart which ensures that the restart is only done once. Is this true? 

Maybe one needs an additional per-hook configuration or property which says if the hook should be called on each target or once after all targets (targets can also be only a single target). The hook will either get a single target or a list of targets as argument.

**Comment by oantolin** [comment]
Yes, currently both `embark--restart` and `embark--confirm` get that type of special treatment. Your magic number is 3, right? If we add an `embark--deselect` too, it would be time to generalize this special treatment. :)

**Comment by minad** [comment]
Lol, right. At least this configuration/property wouldn't have to be public. It could be a private list embark--coalesce-hooks.

**Comment by oantolin** [comment]
If we added a separate `embark-post-act-all-hooks` variable, I think it would weird to not also add `embark-pre-act-all-hooks` and `embark-around-act-all-hooks`. 😬 On the other hand, if we did add those, we could probably remove the `embark-quit-after-action` user variable in favor of putting `embark--confirm` on `embark-pre-act-all-hooks`.

In retrospect maybe it would have been nice to _only_ have the around hooks, which can emulate the others fairly easily. 😃

**Comment by minad** [comment]
On 5/18/23 16:47, Omar Antolín Camarena wrote:
> If we added a separate |embark-post-act-all-hooks| variable, I think it
> would weird to not also add |embark-pre-act-all-hooks| and
> |embark-around-act-all-hooks|. 😬 On the other hand, if we did add
> those, we could probably remove the |embark-quit-after-action| user
> variable in favor of putting |embark--confirm| on
> |embark-pre-act-all-hooks|.

I think I had suggested this before, but my solution would be to specify
the hook type in addition to the hook, such that we don't need these
many customization variables. Then it wouldn't hurt too much if we have
ten different types of hooks. A hypothetical embark-action-hooks
variable could have entries of this form:

(action :pre ... :post ... :pre-all ... :post-all ... :around ...)

> In retrospect maybe it would have been nice to /only/ have the around
> hooks, which can emulate the others fairly easily. 😃

You can still do this and introduce an embark--run-before-and-after
around hook which is added to the :always key such that it is executed
for every action.
